index.js

```diff
@@ -29,50 +29,56 @@ function fastifyEarlyHints (fastify, opts, next) {
       for (const key of Object.keys(headers)) {
         if (Array.isArray(headers[key])) {
           for (const value of headers[key]) {
-            message += `${key}: ${value}${CRLF}`
+            message += `${key}: ${value}${separator}`
           }
         } else {
-          message += `${key}: ${headers[key]}${CRLF}`
+          message += `${key}: ${headers[key]}${separator}`
         }
       }
     } else {
       return Promise.reject(Error(`"headers" expected to be object or Array, but received ${typeof headers}`))
     }
 
-    return new Promise(function (resolve) {
-      if (reply.raw.socket === null) {
-        resolve()
-        return
-      }
-      reply.raw.socket.write(`${EarlyHints}${message}${CRLF}`, 'ascii', () => {
-      // we do not care the message is sent or lost. Since early hints
-      // is metadata to instruct the clients to do something before actual
-      // content. It should never affect the final result if it lost.
-        resolve()
-      })
-    })
+    return write(this, message)
   }
 
   function writeEarlyHintsLinks (links) {
-    const reply = this
     let message = ''
     for (let i = 0; i < links.length; i++) {
-      message += `${formatEntry(links[i], formatEntryOpts)}${CRLF}`
+      message += `${formatEntry(links[i], formatEntryOpts)}${separator}`
     }
 
-    return new Promise(function (resolve) {
-      if (reply.raw.socket === null) {
-        resolve()
-        return
-      }
-      reply.raw.socket.write(`${EarlyHints}${message}${CRLF}`, 'ascii', () => {
-      // we do not care the message is sent or lost. Since early hints
-      // is metadata to instruct the clients to do something before actual
-      // content. It should never affect the final result if it lost.
+    return write(this, message)
+  }
+
+  const write = http2
+    ? function write (reply, message) {
+      return new Promise(function (resolve) {
+        if (reply.raw.socket === null) {
```

Comment:
Is this also valid for http2 connections?

Reply:
`http2` should be hardly hit this since it share the same `socket` for multiple `stream`

Reply:
So we can keep it, i guess.